packages/grid/_modules_/grid/hooks/features/virtualization/useGridVirtualRows.ts

```diff
@@ -144,6 +149,17 @@ export const useGridVirtualRows = (apiRef: GridApiRef): void => {
         top: containerProps.isVirtualized ? rzScrollTop : scrollTop,
       };
 
+      if (page !== nextPage) {
+        const newPage = page < nextPage ? nextPage + 1 : nextPage;
+        const nextPageFirstRowIndex = newPage * containerProps.viewportPageSize;
```

Comment:
instead of recalculating the first row index, you can get it from the rendering state `state.rendering`

Reply:
which you can just pass as argument

Reply:
The thing is that I only have access to the current rendering state and I need to send the next rendering state values, not the current `page` but the `nextPage`. Even if I used the values from the `state.rendering` I would still need to calculate them. Also, the scroll direction (up or down) makes a difference.

Reply:
you should  trigger the event after the state is updated

Reply:
This will be removed altogether as it will be decoupled from the virtualization.

Reply:
Using the state that determines what to render is an interesting idea (proposed by @dtassone), with a high potential. I mean, conceptually speaking, the virtualization needs to know exactly the same information as we want to fire in the event (plus the overscan).

Sure, the virtualization currently doesn't compute this state right: #1911, but it will need to. In react-virtualized, the component uses this information under [onRowsRendered](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md) with the viewport start and end, as well as the overscan. It's then [used by](https://github.com/bvaughn/react-virtualized/blob/abe0530a512639c042e74009fbf647abdb52d661/source/InfiniteLoader/InfiniteLoader.js#L119) the `InfiniteLoader` to have a rich loading logic.

We get the same API in react-window, under [onItemsRendered](https://react-window.vercel.app/#/api/FixedSizeList). 

Looking at the source of mui-plus, it seems to be done right: https://github.com/Janpot/mui-plus/blob/11a4cb121e7e3d3ad0b5c1190e23927313a9abc3/packages/mui-plus/src/utils/virtualization.ts#L69. It also relevant to mention that for the more advanced use cases, variable height and dynamic height, it's doesn't seem realistic to duplicate the logic.

Reply:
Updated